miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,813 @@
+            assert!(is_set);
+            assert_eq!(is_set, sparse_nodes.get(idx).unwrap().is_none());
+
+            let &test_node = sparse_nodes.get(idx).unwrap();
+            assert_eq!(test_node, None);
+        }
+
+        // Depth 1.
+        {
+            let depth: u8 = 1;
+            let idx = (sparse_path.depth() - depth) as usize;
+            assert_eq!(idx, 7);
+            let bit = 0b0000_0001;
+            assert_eq!(bit, 1 << (depth - 1));
+            let is_set = (sparse_path.empty_nodes_mask & bit) != 0;
+            assert!(is_set);
+            assert_eq!(is_set, sparse_nodes.get(idx).unwrap().is_none());
+
+            let &test_node = sparse_nodes.get(idx).unwrap();
+            assert_eq!(test_node, None);
+        }
+    }
+
+    #[test]
+    fn from_sized_iter() {
+        let tree = make_smt(8192);
+
+        for (key, _value) in tree.entries() {
+            let index = NodeIndex::from(Smt::key_to_leaf_index(key));
+
+            let control_path = tree.get_path(key);
+            for (&control_node, proof_index) in iter::zip(&*control_path, index.proof_indices()) {
+                let proof_node = tree.get_hash(proof_index);
+                assert_eq!(control_node, proof_node, "WHat");
+            }
+
+            let sparse_path =
+                SparseMerklePath::from_sized_iter(control_path.clone().into_iter()).unwrap();
+            for (sparse_node, proof_idx) in iter::zip(sparse_path.clone(), index.proof_indices()) {
+                let proof_node = tree.get_hash(proof_idx);
+                assert_eq!(sparse_node, proof_node, "WHat");
+            }
+
+            assert_eq!(control_path.depth(), sparse_path.depth());
+            for (i, (control, sparse)) in iter::zip(control_path, sparse_path).enumerate() {
+                assert_eq!(control, sparse, "on iteration {i}");
+            }
+        }
+    }
+
+    #[test]
+    fn test_random_access() {
+        let tree = make_smt(8192);
+
+        for (i, (key, _value)) in tree.entries().enumerate() {
+            let control_path = tree.get_path(key);
+            let sparse_path = SparseMerklePath::try_from(control_path.clone()).unwrap();
+            assert_eq!(control_path.depth(), sparse_path.depth());
+            assert_eq!(sparse_path.depth(), SMT_DEPTH);
+
+            // Test random access by depth.
+            for depth in path_depth_iter(control_path.depth()) {
+                let &control_node = control_path.at_depth(depth).unwrap();
+                let sparse_node = sparse_path.at_depth(depth).unwrap();
+                assert_eq!(control_node, sparse_node, "at depth {depth} for entry {i}");
+            }
+
+            // Test random access by index.
+            // Letting index get to `control_path.len()` will test that both sides correctly return
+            // `None` for out of bounds access.
+            for index in 0..=(control_path.len()) {
+                let control_node = control_path.at_idx(index).copied();
```

Comment:
Nit: Maybe `MerklePath::at_idx` should return `Option<RpoDigest>`, i.e. do the `copied()` internally?